src/rdkafka_metadata.c

```diff
@@ -954,17 +1001,17 @@ rd_kafka_parse_Metadata0(rd_kafka_broker_t *rkb,
                            "Caching full metadata with "
                            "%d broker(s) and %d topic(s): %s",
                            md->broker_cnt, md->topic_cnt, reason);
-        } else {
-                if (cache_changes)
-                        rd_kafka_metadata_cache_propagate_changes(rk);
-                rd_kafka_metadata_cache_expiry_start(rk);
         }
-
-
-        // TODO: Should be done for requested_topic_ids as well.
         /* Remove cache hints for the originally requested topics. */
         if (requested_topics)
                 rd_kafka_metadata_cache_purge_hints(rk, requested_topics);
+        if (requested_topic_ids)
+                rd_kafka_metadata_cache_purge_hints(rk, requested_topic_ids);
+
+        if (cache_changes) {
+                rd_kafka_metadata_cache_propagate_changes(rk);
+                rd_kafka_metadata_cache_expiry_start(rk);
```

Comment:
- Do you mean to include expiry start call inside the if condition? Earlier it was outside.
- What happens to the cache entries which were not updated? Do we restart the timer for them as well?

Reply:
Here it's restarting the timer if some entries have been updated, the timer is calculated from the expiry time of the first element, the one that expires first. Earlier was outside just in this code, in trunk it's done only if `!all_topics` but now existing entries can be updated in that case.